### ISSUE TITLE
Create used bean without injection

### ISSUE THREAD

**Issue by chris922** [mapstruct/mapstruct#1497]
As discussed on Gitter with @sjaakd and @filiphr it would be great to have the possibility to determine if a referenced bean/mapper should be "created" based on the chosen component model or by instantiating it manually.

The idea is to extend the `@Mapper` annotation with a new attribute `usesPlain` which will ensure that the given classes will be instantiated and not injected.

Example:

```java
@Mapper(uses = BeanMapper.class, usesPlain = ObjectFactory.class, componentModel = "spring")
interface FoobarMapper {
  // ...
}

// generated
@Component
public class FoobarMapperImpl implements FoobarMapper {
  @Autowired
  private BeanMapper beanMapper;

  private ObjectFactory objectFactory = new ObjectFactory();

  // ...
}
```

**Comment by sjaakd** [comment]
And.. this new `usesPlain` should only be possible icw a component model.. (Spring, JSR303, CDI)

**Comment by chris922** [comment]
So you mean it should be ignored in case a `default` component model will be used? Or an error reported?

I would expect that it will still work. Don't think it will make troubles, but simplifies it when someone switches between component models (or the component model was defined somewhere else and the mapper itself doesn't know which model is active)

**Comment by sjaakd** [comment]
> So you mean it should be ignored in case a default component model will be used? Or an error reported?

If I followed the discussion on Gitter, @filiphr  wanted to raise an error in this case.. But perhaps I'm mistaken..

**Comment by filiphr** [comment]
I don't think that it needs to be ignored and definitely we don't need to raise an error.

We can output a warning to the user so they know that they are trying to use things which make no sense.

However, I think that we should error out if the same type is used both in `uses` and `usesPlain`

**Comment by gunnarmorling** [comment]
@chris922 Out of curiousity, what is the use case for this? It seems you always could make such object injectable, e.g. by creating a matching producer method in CDI or by declaring a bean of that type in Spring. On first thought, this seems to add a bit complexity and I'm wondering what it gains us?

**Comment by chris922** [comment]
In our project we've got > 100 generated XML object factories.. defining everyone manually as a bean in Spring would therefore make it really complex and it's a lot of boilerplate for this configuration.. and it will reduce startup time as a lot more beans must be registered.

**Comment by sjaakd** [comment]
> All in all, I'm not a huge fan of this feature, as it adds complexity to
> the API surface and already can be addressed using existing means. Not sure
> what the others think?

Well, one of the problems that we face is how to recognise an object factory. @agudian introduced the `@ObjectFactory` annotation. An object annotated with `@ObjectFactory` produces a target, but can take zero or more source (or `@Context` annotated) arguments as well. This in contrast to the 'old' style factories (once implemented by me) that only return a target object because otherwise they could be mixed up with mapping methods.

There's the fact that you cannot apply `@ObjectFactory` annotation to generated (e.g. JAXB xmlrepositories) factories. This is the more or less the same problem that @chris922 is facing (although it can probably be addressed with an `@Produces`). However, you sometimes need to provide a source argument (`JAXBELement`) to produce the proper target. 

Since the problem is mainly related to object factories, I would propose to have a new argument (as 'usePlain' but then called 'legacyObjectFactories'). I would no longer allow to add no-arg, non-annotated object factories to 'uses' (and warn the user to use 'legacyObjectFactory' from now on). Advantage: you have a similar treatment / capabilities for object factories with or without the annotation. They would be treated internally the same.

**Comment by gunnarmorling** [comment]
@sjaakd I have to admit I don't fully grok what you're proposing. Could you perhaps add a bit of example code that shows how this would be used?

**Comment by sjaakd** [comment]
This is generated code from JAXB:

```java
@XmlRegistry
public class ObjectFactory {

    private final static QName _C_QNAME = new QName("c", "c");
    private final static QName _C2_QNAME = new QName("c", "c2");

    /**
     * Create a new ObjectFactory that can be used to create new instances of schema derived classes for package: c
     * 
     */
    public ObjectFactory() {
    }

    /**
     * Create an instance of {@link C2Type }
     * 
     */
    public C2Type createC2Type() {
        return new C2Type();
    }

    /**
     * Create an instance of {@link CType }
     * 
     */
    public CType createCType() {
        return new CType();
    }

    /**
     * Create an instance of {@link C1Type }
     * 
     */
    public C1Type createC1Type() {
        return new C1Type();
    }

    /**
     * Create an instance of {@link JAXBElement }{@code <}{@link CType }{@code >}}
     * 
     */
    @XmlElementDecl(namespace = "c", name = "c")
    public JAXBElement<CType> createC(CType value) {
        return new JAXBElement<CType>(_C_QNAME, CType.class, null, value);
    }

    /**
     * Create an instance of {@link JAXBElement }{@code <}{@link C2Type }{@code >}}
     * 
     */
    @XmlElementDecl(namespace = "c", name = "c2")
    public JAXBElement<C2Type> createC2(C2Type value) {
        return new JAXBElement<C2Type>(_C2_QNAME, C2Type.class, null, value);
    }

}
```

Problem 1: The methods `public JAXBElement<CType> createC(CType value)` and `public JAXBElement<C2Type> createC2(C2Type value) ` are **not** selected as factory method by MapStruct. However they would be selected if they were annotated with `@ObjectFactory` which is impossible since they are generated. 

Problem 2: This issue. Using a component model, they are **not** injected (you could perhaps use the `@Produces` as you pointed out). This requires explanation, therefore I propose to  use  'new'.

Now: @chris922 proposes to have a `@Mapper#usesPlain`: applicable to mappers and object factories. I propose to limit it to object factories only: `@Mapper#legacyObjectFactory`.  IMHO it is an object factory problem not a mapper problem. 

Solution is simple: make behaviour of generated (legacy) object factories equal to behaviour of `@ObjectFactory` annotated object factories. That requires object factories to be distinct from mappers. Hence `@Mapper#legacyObjectFactories` and asking the user to move object factories from  `@Mapper#uses`  to  `@Mapper#legacyObjectFactories`.

**Comment by sjaakd** [comment]
So.. as we discussed in our hangout. @agudian noted correctly that there are 2 aspects, which should be treated separately. We came up with the idea of using meta annotations.

This could look like this:

```java
@Mapper(uses = { BeanMapper.class, JaxbObjectFactories.class} , componentModel = "spring")
interface FoobarMapper {
  // ...
}

@UsedObjectFactories( { org.something.jaxb.ObjectFactory.class }, useNoComponentModelToInstantiate = true ) 
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.CLASS)
public @interface JaxbObjectFactories.class {
}

```

So.. you can create your own groups of object factories with certain properties. Moreover, you could extend this with another annotation like `UsedMappers` when the need arises (Mappers is already taken, hence the name `UsedObjectFactories`. All the methods inside an object factory class are treated as if they would have the `@ObjectFactory` annotations.

Remaining point of debate: should we break backwards compatibility, making the behaviour of all object-factory-methods the same and signal the user(error)  in 1.3 or 1.4 to go to the newer ways of defining (`@ObjectFactory` or `@UsedObjectFactories`)?

1. maintain backward compatibility
2. break backward compatibility gracefully for the sake of consistent api behaviour

@chris922 , @agudian , @filiphr , @gunnarmorling  could you give an opinion on:
a. the overal solution
b. 1. or 2. ?

As a side track: the `@UsedMapper` annotation might be a way to do something to help on #1509  `UsedMapper` having a reusable single point of defining all mappers (I definitely prefer defining over scanning)

**Comment by chris922** [comment]
Hum.. looks somehow complicated to me. So you will reference to a meta-annotation in `uses` instead of referencing to the required mapper itself... it acts like a "placeholder".

I got it in our hangout more in a way like this:

```java
@Mapper(uses = {SpringBeanMapper.class, JaxbObjectFactory.class}, componentModel = "spring")
@UsedMapperConfig(mapper = JaxbObjectFactory.class, componentModel = "none")
interface FoobarMapper { ... }
```

.. but having a way to define a group of mappers is also  interesting. And as every mapper will most likely always be used with the same component model it makes sense to have the configuration not on each mapper but somewhere centralized .. 

Towards your question `b`: In general I would always say maintain backward compatibility in minor releases .. but this doesn't affect already generated mappers in libraries you are not touching and you will directly notice that you need to change something when you press 'generate' .. hum .. hard to decide what is the best way to go without making an upgrade to a newer version to hard for the users.

**Comment by sjaakd** [comment]
> @Mapper(uses = {SpringBeanMapper.class, JaxbObjectFactory.class}, componentModel = "spring")
> @UsedMapperConfig(mapper = JaxbObjectFactory.class, componentModel = "none")
> interface FoobarMapper { ... }

Yes.. that was indeed the original idea. However, the drawback is that you have to refer to the `JaxbObjectFactory` twice and further refine it by the `@UsedMapperConfig`. The concept above is something we already have in the form of qualifiers. The general idea (for the future) is that you you have 'pieces' of reusable configuration that you can re-use or even pre-define as MapStruct annotations. If you don't want to do that, just refer directly.

**Comment by chris922** [comment]
At least I like the part of having reusable configuration, so yes - maybe that's the way it should be done.

@gunnarmorling what do you think? You had some concerns about the first solution (this PR) - what about the new solution @sjaakd proposed?
